File: leetcode/347TopKFrequentElements/topKelements.py
from collections import defaultdict
from typing import List

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq: List[int, int] = defaultdict(int)
        bucket: List[List[int]] = [[] for _ in range(len(nums) + 1)]

        for num in nums:
            freq[num] += 1
        for key, val in freq.items():
            bucket[val].append(key)
        
        result: List[int] = []
        for i in reversed(range(len(nums) + 1)):
            for j in bucket[i]:
                result.append(j)
                if len(result) == k:
                    return result

if __name__ == "__main__":
    sol = Solution()
    input = [[1,1,1,2,2,3], 2]
    output = sol.topKFrequent(*input)
    print(output)

File: leetcode/0347TopKFrequentElements/topKelements.cpp
#include <vector>
#include <unordered_map>
#include <iostream>
#include <string>

class Solution {
public:
    std::vector<int> topKFrequent(std::vector<int>& nums, int k) {
        std::unordered_map<int, int> freq;
        std::vector<std::vector<int>> bucket(nums.size() + 1);

        for (const int& n : nums) {
            freq[n]++;
        }
        for (const auto& pair : freq) {
            bucket[pair.second].push_back(pair.first);
        }

        std::vector<int> result;
        for (int i = nums.size(); i > 0; i--) {
            for (const int& n : bucket[i]) {
                result.push_back(n);
                if (result.size() == k) return result;
            }
        }

        return {};
    }
};

int main() {
    Solution sol = Solution();
    std::vector<int> nums = {1,1,1,2,2,3};
    std::vector<int> res = sol.topKFrequent(nums, 2);
    for (const int& n : res) {
        std::cout << std::to_string(n) << std::endl; 
    }
}